# Week 3 Assignment

In this part, I am reading the data from the Wikipedia page and storing it in 'df'

In [20]:
import pandas as pd

In [41]:
dfs = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M',header=0)
df = dfs[0]

In [42]:
df.columns=['PostalCode', 'Borough', 'Neighbourhood']
df.head()

PostalCode           Borough     Neighbourhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [43]:
df.shape

(288, 3)

As can be seen, there are 'Not Assigned' Values in columns 'Borough' and 'Neighborhood'.

### Removing 'Not Assigned' values from the data frame

I am removing 'Not Assigned' values from the column 'Borough' using the following command:

In [44]:
df= df[df['Borough'] != 'Not assigned']

In [45]:
df.shape

(211, 3)

As can be seen, the number of rows is decreased. It was 288 and after removing 'Not Assigned' values, there is only 211 rows in the data frame.

In [46]:
df.head(10)

PostalCode           Borough     Neighbourhood
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M5A  Downtown Toronto       Regent Park
6         M6A        North York  Lawrence Heights
7         M6A        North York    Lawrence Manor
8         M7A      Queen's Park      Not assigned
10        M9A         Etobicoke  Islington Avenue
11        M1B       Scarborough             Rouge
12        M1B       Scarborough           Malvern

There is only one row that the neighborhood is not assigned

In [29]:
#df.iloc[8, 2] = 'Queen\'s Park'

In [38]:
#df.head(40)

As can be seen, there are no 'Not Assigned' values in neither 'Borough' nor 'Neighbourhood'

### Adding Postal Code to the data frame

Grouping the neighborhood based on their postral code

In [47]:

df_post = df.groupby(['PostalCode']).agg({'Borough':pd.Series.unique,'Neighbourhood':lambda x:','.join(x)})

In [48]:
df_post.reset_index(inplace = True)

In [49]:
df.head()

PostalCode           Borough     Neighbourhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M5A  Downtown Toronto       Regent Park
6        M6A        North York  Lawrence Heights

In [51]:
df_post.shape

(103, 3)

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [53]:

df_post.loc[df_post['Neighbourhood'] =='Not assigned','Neighbourhood'] = df_post.loc[df_post['Neighbourhood'] =='Not assigned', 'Borough']

In [54]:
df_post.shape

(103, 3)

# Second Part 

In [56]:
geo = pd.read_csv('http://cocl.us/Geospatial_data', header = 0 )

In [57]:
geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Changing 'Postal Code' to 'PostalCode' in order to be compatible with the previous dataset (df)

In [58]:

geo.rename(columns={"Postal Code":"PostalCode"}, inplace = True)

In [59]:
geo.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

As can be seen, the name of the column has changed. The following command will merge the two datasets based on the Postal Code.

In [60]:
df = pd.merge(df_post, geo, on='PostalCode')
df.shape

(103, 5)

Now, we'll a few rown of the new dataset :

In [61]:
df.head(10)

PostalCode      Borough                                  Neighbourhood  \
0        M1B  Scarborough                                  Rouge,Malvern   
1        M1C  Scarborough           Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough                Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                         Woburn   
4        M1H  Scarborough                                      Cedarbrae   
5        M1J  Scarborough                            Scarborough Village   
6        M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park   
7        M1L  Scarborough                  Clairlea,Golden Mile,Oakridge   
8        M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West   
9        M1N  Scarborough                     Birch Cliff,Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

In [35]:
df.shape

(211, 3)

# Third Part

Generating map of toronto and clustering the different neighbourhoods. First, we need to import the necessary libraries.

In [62]:
from geopy.geocoders import Nominatim

import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

Based on the labs, we need to define the address and location

In [63]:
address = 'Toronto'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.653963, -79.387207.


In [64]:
# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

In [66]:
df_Toronto = df[df['Borough'].str.contains('Toronto')].reset_index(drop=True)
df_Toronto

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                         The Danforth West,Riverdale  43.679557 -79.352188  
2                       The Beaches West,India Bazaar  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                                  North Toronto West  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                          Moore Park,Summerhill East  43.689574 -79.383160  
9   Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                         Cabbagetown,St. James Town  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                           Harbourfront,Regent Park  43.654260 -79.360636  
14                            Ryerson,Garden District  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                             Adelaide,King,Richmond  43.650571 -79.384568  
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
24                  The Annex,North Midtown,Yorkville  43.672710 -79.405678  
25                      Harbord,University of Toronto  43.662696 -79.400049  
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
29              First Canadian Place,Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                        Dovercourt Village,Dufferin  43.669005 -79.442259  
32                            Little Portugal,Trinity  43.64792

In [67]:
# create map of Toronto Borough using latitude and longitude values
map_Toronto_Borough = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_Toronto['Latitude'], df_Toronto['Longitude'], df_Toronto['Borough'], df_Toronto['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto_Borough)  
    
map_Toronto_Borough

We need our ID and secret from foursquare developer account.

In [68]:
CLIENT_ID = 'NCJ0VG54N3HI1VTDLYFVWAR2SOHO5FFM0MKB50V4PNPKBN4G'
CLIENT_SECRET = '5EIKV3ACBYW1QEWSQSAWCWKNU2WVCXI13HFTFCP4XNRMSMVC'
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: NCJ0VG54N3HI1VTDLYFVWAR2SOHO5FFM0MKB50V4PNPKBN4G
CLIENT_SECRET:5EIKV3ACBYW1QEWSQSAWCWKNU2WVCXI13HFTFCP4XNRMSMVC


In [69]:
df_Toronto.shape

(38, 5)

In [70]:
df_Toronto.loc[0, 'Neighbourhood']

'The Beaches'

In [71]:
#Get the neighborhood's latitude and longitude values.
neighbourhood_latitude = df_Toronto.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = df_Toronto.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = df_Toronto.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [72]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=NCJ0VG54N3HI1VTDLYFVWAR2SOHO5FFM0MKB50V4PNPKBN4G&client_secret=5EIKV3ACBYW1QEWSQSAWCWKNU2WVCXI13HFTFCP4XNRMSMVC&v=20180604&ll=43.67635739999999,-79.2930312&radius=500&limit=100'

In [73]:
#Send the GET request and examine the resutls
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d225425787dba00384a6e1f'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bd461bc77b29c74a07d9282',
       'name': 'Glen Manor Ravine',
       'location': {'address': 'Glen Manor',
        'crossStreet': 'Queen St.',
        'lat': 43.67682094413784,
        'lng': -79.29394208780985,
        'labeledLatLngs': [{'labe

In [74]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [75]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name            categories        lat  \
0                   Glen Manor Ravine                 Trail  43.676821   
1  The Big Carrot Natural Food Market     Health Food Store  43.678879   
2                 Grover Pub and Grub                   Pub  43.679181   
3                 Glen Stewart Ravine  Other Great Outdoors  43.676300   
4                       Upper Beaches          Neighborhood  43.680563   

         lng  
0 -79.293942  
1 -79.297734  
2 -79.297215  
3 -79.294784  
4 -79.292869

In [76]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

5 venues were returned by Foursquare.


In [77]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [81]:
Toronto_venues = getNearbyVenues(names=df_Toronto['Neighbourhood'],
                                   latitudes=df_Toronto['Latitude'],
                                   longitudes=df_Toronto['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvall

In [83]:

#Let's check the size of the resulting dataframe
print(Toronto_venues.shape)
Toronto_venues.head()

(1700, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0   The Beaches               43.676357               -79.293031   
1   The Beaches               43.676357               -79.293031   
2   The Beaches               43.676357               -79.293031   
3   The Beaches               43.676357               -79.293031   
4   The Beaches               43.676357               -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                 Glen Stewart Ravine       43.676300       -79.294784   
4                       Upper Beaches       43.680563       -79.292869   

         Venue Category  
0                 Trail  
1     Health Food Store  
2                   Pub  
3  Other Great Outdoors  
4          Neighborhood

In [84]:
#Let's check how many venues were returned for each neighborhood
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             55   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      16   
Cabbagetown,St. James Town                                              46   
Central Bay Street                                                      88   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                15   
Church and Wellesley                                                    87   
Commerce Court,Victoria Hotel                                          100   
Davisville                                                              38   
Davisville North                                                         8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                      14   
Design Exchange,Toronto Dominion Centre                                100   
Dovercourt Village,Dufferin                                             20   
First Canadian Place,Underground city                                  100   
Forest Hill North,Forest Hill West                                       4   
Harbord,University of Toronto                                           34   
Harbourfront East,Toronto Islands,Union Station                        100   
Harbourfront,Regent Park                                                48   
High Park,The Junction South                                            23   
Lawrence Park                                                            3   
Little Portugal,Trinity                                                 62   
Moore Park,Summerhill East                                               2   
North Toronto West                                                      16   
Parkdale,Roncesvalles                                                   15   
Rosedale                                                                 5   
Roselawn                                                                 2   
Runnymede,Swansea                                                       38   
Ryerson,Garden District                                                100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         94   
Studio District                                                         38   
The Annex,North Midtown,Yorkville                                       23   
The Beaches                                                              5   
The Beaches West,India Bazaar                                           18   
The Danforth West,Riverdale                                             42   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide,King,Richmond                                                  100   
Berczy Park                                                              55   
Brockton,Exhibition Place,Parkdale Village                               22   
Business Reply Mail Processing Centre 969 Eastern                        19   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                       16   
Cabbagetown,St. James Town                                               46   
Central Bay Street                                                       88   
Chinatown,Grange Park,Kensington Market                                 100   
Christie                                                                 15   
Church and Wellesley                                          

In [85]:

# Let's find out how many unique categories can be curated from all the returned venues
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 239 uniques categories.


In [86]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head(100)

Neighbourhood  Adult Boutique  Afghan Restaurant  Airport  \
0                     The Beaches               0                  0        0   
1                     The Beaches               0                  0        0   
2                     The Beaches               0                  0        0   
3                     The Beaches               0                  0        0   
4                     The Beaches               0                  0        0   
5     The Danforth West,Riverdale               0                  0        0   
6     The Danforth West,Riverdale               0                  0        0   
7     The Danforth West,Riverdale               0                  0        0   
8     The Danforth West,Riverdale               0                  0        0   
9     The Danforth West,Riverdale               0                  0        0   
10    The Danforth West,Riverdale               0                  0        0   
11    The Danforth West,Riverdale               0                  0        0   
12    The Danforth West,Riverdale               0                  0        0   
13    The Danforth West,Riverdale               0                  0        0   
14    The Danforth West,Riverdale               0                  0        0   
15    The Danforth West,Riverdale               0                  0        0   
16    The Danforth West,Riverdale               0                  0        0   
17    The Danforth West,Riverdale               0                  0        0   
18    The Danforth West,Riverdale               0                  0        0   
19    The Danforth West,Riverdale               0                  0        0   
20    The Danforth West,Riverdale               0                  0        0   
21    The Danforth West,Riverdale               0                  0        0   
22    The Danforth West,Riverdale               0                  0        0   
23    The Danforth West,Riverdale               0                  0        0   
24    The Danforth West,Riverdale               0                  0        0   
25    The Danforth West,Riverdale               0                  0        0   
26    The Danforth West,Riverdale               0                  0        0   
27    The Danforth West,Riverdale               0                  0        0   
28    The Danforth West,Riverdale               0                  0        0   
29    The Danforth West,Riverdale               0                  0        0   
30    The Danforth West,Riverdale               0                  0        0   
31    The Danforth West,Riverdale               0                  0        0   
32    The Danforth West,Riverdale               0                  0        0   
33    The Danforth West,Riverdale               0                  0        0   
34    The Danforth West,Riverdale               0                  0        0   
35    The Danforth West,Riverdale               0                  0        0   
36    The Danforth West,Riverdale               0                  0        0   
37    The Danforth West,Riverdale               0                  0        0   
38    The Danforth West,Riverdale               0                  0        0   
39    The Danforth West,Riverdale               0                  0        0   
40    The Danforth West,Riverdale               0                  0        0   
41    The Danforth West,Riverdale               0                  0        0   
42    The Danforth West,Riverdale               0                  0        0   
43    The Danforth West,Riverdale               0                  0        0   
44    The Danforth West,Riverdale               0                  0        0   
45    The Danforth West,Riverdale               0                  0        0   
46    The Danforth West,Riverdale               0                  0        0   
47  The Beaches West,India Bazaar               0                  0        0   
48  The Beaches West,India Bazaar               0  

In [87]:
Toronto_onehot.shape

(1700, 240)

In [89]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped.head()

Neighbourhood  Adult Boutique  \
0                             Adelaide,King,Richmond             0.0   
1                                        Berczy Park             0.0   
2         Brockton,Exhibition Place,Parkdale Village             0.0   
3  Business Reply Mail Processing Centre 969 Eastern             0.0   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...             0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0   0.0000              0.0000        0.0000   
1                0.0   0.0000              0.0000        0.0000   
2                0.0   0.0000              0.0000        0.0000   
3                0.0   0.0000              0.0000        0.0000   
4                0.0   0.0625              0.0625        0.0625   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0           0.000            0.000             0.000                 0.04   
1           0.000            0.000             0.000                 0.00   
2           0.000            0.000             0.000                 0.00   
3           0.000            0.000             0.000                 0.00   
4           0.125            0.125             0.125                 0.00   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0     0.010000        0.01                  0.0   
1           0.0       0.0     0.018182        0.00                  0.0   
2           0.0       0.0     0.000000        0.00                  0.0   
3           0.0       0.0     0.000000        0.00                  0.0   
4           0.0       0.0     0.000000        0.00                  0.0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0              0.02       0.000000        0.0         0.0    0.000000   
1              0.00       0.000000        0.0         0.0    0.018182   
2              0.00       0.000000        0.0         0.0    0.000000   
3              0.00       0.052632        0.0         0.0    0.000000   
4              0.00       0.000000        0.0         0.0    0.000000   

     Bakery  Bank       Bar  Baseball Stadium  Basketball Stadium     Beach  \
0  0.020000   0.0  0.040000               0.0            0.000000  0.000000   
1  0.036364   0.0  0.000000               0.0            0.018182  0.018182   
2  0.045455   0.0  0.045455               0.0            0.000000  0.000000   
3  0.000000   0.0  0.000000               0.0            0.000000  0.000000   
4  0.000000   0.0  0.062500               0.0            0.000000  0.000000   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant    Bistro  \
0              0.0  0.000000         0.0            0.000000  0.000000   
1              0.0  0.036364         0.0            0.018182  0.018182   
2              0.0  0.000000         0.0            0.000000  0.000000   
3              0.0  0.000000         0.0            0.000000  0.000000   
4              0.0  0.000000         0.0            0.000000  0.000000   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0         0.0000       0.01    0.0000                  0.01        0.030000   
1         0.0000       0.00    0.0000                  0.00        0.018182   
2         0.0000       0.00    0.0000                  0.00        0.090909   
3         0.0000       0.00    0.0000                  0.00        0.000000   
4         0.0625       0.00    0.0625                  0.00        0.000000   

    Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  0.000000              0.0       0.0          0.03       0.010000       0.0   
1  0.000000              0.0       0.0          0.00       0.000000       0.0   
2  0.000000              0.0       0.0          0.00       0.045455       0.0   
3  0.052632              0.0       0.0          0.00       0.052632       0.0   
4  0.000000              0.0       0.0          0.

In [90]:

Toronto_grouped.shape

(38, 240)

In [91]:
num_top_venues = 5

for hood in Toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2                  Bar  0.04
3      Thai Restaurant  0.04
4  American Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2          Steakhouse  0.04
3  Italian Restaurant  0.04
4                Café  0.04


----Brockton,Exhibition Place,Parkdale Village----
               venue  freq
0     Breakfast Spot  0.09
1        Coffee Shop  0.09
2               Café  0.09
3             Bakery  0.05
4  Convenience Store  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0         Yoga Studio  0.05
1          Smoke Shop  0.05
2       Garden Center  0.05
3              Garden  0.05
4  Light Rail Station  0.05


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge  0.12

In [92]:
# First, let's write a function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [93]:
# Now let's create the new dataframe and display the top 10 venues for each neighborhood.
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant                   Bar   
1          Cocktail Bar                  Café              Beer Bar   
2        Breakfast Spot                  Café           Yoga Studio   
3           Pizza Place               Brewery            Skate Park   
4       Airport Service      Airport Terminal              Boutique   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse    American Restaurant        Cosmetics Shop   
1            Steakhouse                 Bakery    Seafood Restaurant   
2         Burrito Place  Performing Arts Venue               Stadium   
3         Burrito Place                Butcher            Restaurant   
4       Harbor / Marina       Sculpture Garden         Boat or Ferry   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Burger Joint            Restaurant                  Hotel  
1        Farmers Market           Cheese Shop     Italian Restaurant  
2  Caribbean Restaurant            Restaurant                    Bar  
3      Recording Studio            Comic Shop                    Spa  
4                   Bar                 Plane            Coffee Shop

In [94]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [95]:
# Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = df_Toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

Toronto_merged.head() # check the last columns!

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                   Pub  Other Great Outdoors   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park    Italian Restaurant   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               3                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0     Health Food Store          Neighborhood                   Trail   
1    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
2             Pet Store           Pizza Place                     Pub   
3             Gastropub    Italian Restaurant                  Bakery   
4              Bus Line           Yoga Studio                 Dog Run   

  6th Most Common Venue 7th Most Common Venue    8th Most Common Venue  \
0               Dog Run      Doner Restaurant               Donut Shop   
1               Brewery                Bakery  Fruit & Vegetable Store   
2         Movie Theater        Sandwich Place            Burrito Place   
3   American Restaurant           Yoga Studio     Gym / Fitness Center   
4   Filipino Restaurant  Fast Food Restaurant           Farmers Market   

  9th Most Common Venue       10th Most Common Venue  
0   Dumpling Restaurant  Eastern European Restaurant  
1             Juice Bar                 Liquor Store  
2          Burger Joint                      Brewery  
3               Brewery           Seafood Restaurant  
4    Falafel Restaurant                  Event Space

In [96]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [97]:

Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0                   Pub   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0                  Park   
3       East Toronto               0                  Café   
5    Central Toronto               0                  Park   
6    Central Toronto               0           Coffee Shop   
7    Central Toronto               0           Pizza Place   
9    Central Toronto               0                   Pub   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0   Japanese Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0                  Café   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
24   Central Toronto               0           Coffee Shop   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
27  Downtown Toronto               0        Airport Lounge   
28  Downtown Toronto               0           Coffee Shop   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0           Coffee Shop   
34      West Toronto               0                  Café   
35      West Toronto               0        Breakfast Spot   
36      West Toronto               0           Pizza Place   
37      East Toronto               0           Yoga Studio   

            2nd Most Common Venue 3rd Most Common Venue  \
0            Other Great Outdoors     Health Food Store   
1                     Coffee Shop    Italian Restaurant   
2              Italian Restaurant             Pet Store   
3                     Coffee Shop             Gastropub   
5                           Hotel        Clothing Store   
6                     Yoga Studio            Bagel Shop   
7                    Dessert Shop        Sandwich Place   
9                     Coffee Shop          Liquor Store   
11                     Restaurant                  Park   
12                    Coffee Shop      Sushi Restaurant   
13                           Park                Bakery   
14                 Clothing Store        Cosmetics Shop   
15                    Coffee Shop                 Hotel   
16                   Cocktail Bar                  Café   
17             Italian Restaurant                  Café   
18                           Café       Thai Restaurant   
19                       Aquarium                 Hotel   
20                           Café                 Hotel   
21                           Café                 Hotel   
24                           Café        Sandwich Place   
25             Italian Restaurant   Japanese Restaurant   
26  Vegetarian / Vegan Restaurant    Mexican Restaurant   
27                Airport Service      Airport Terminal   
28                           Café            Restaurant   
29                           Café            Restaurant   
30                           Café                  Park   
31                    Supermarket                Bakery   
32               Asian Restaurant           Men's Store   
33                 Breakfast Spot                  Café   
34                            Bar    Mexican Restaurant   
35                      Gift Shop             Bookstore   
36              

In [98]:

Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto               1          Tennis Court   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8            Playground          Concert Hall               Dog Run   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8        Farmers Market    Falafel Restaurant            Event Space

In [99]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               2        Ice Cream Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22                Garden           Yoga Studio               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22        Farmers Market    Falafel Restaurant            Event Space

In [100]:

Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4  Central Toronto               3                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4           Swim School              Bus Line           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4               Dog Run   Filipino Restaurant  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4        Farmers Market    Falafel Restaurant            Event Space

In [101]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto               4                  Park   
23   Central Toronto               4         Jewelry Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10            Playground              Building                 Trail   
23                 Trail                  Park      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10        Discount Store  Fast Food Restaurant        Farmers Market   
23      Doner Restaurant            Donut Shop   Dumpling Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
10           Falafel Restaurant           Event Space   Ethiopian Restaurant  
23  Eastern European Restaurant     Electronics Store            Yoga Studio